react/src/Components/Footer/Components/SpeedTestDialog.js

```diff
@@ -0,0 +1,155 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import DialogTitle from '@mui/material/DialogTitle';
+import Dialog from '@mui/material/Dialog';
+import Box from '@mui/material/Box';
+import Button from '@mui/material/Button';
+import DialogContent from '@mui/material/DialogContent';
+import InputLabel from '@mui/material/InputLabel';
+import Select from '@mui/material/Select';
+import { MediaSettingsContext } from '../../../pages/AntMedia';
+import { AntmediaContext } from '../../../App';
+import {Checkbox, Grid, Hidden, MenuItem, useMediaQuery} from '@mui/material';
+import { SvgIcon } from 'Components/SvgIcon';
+import { useTheme } from '@mui/material';
+import { useTranslation } from 'react-i18next';
+
+const AntDialogTitle = props => {
+  const { children, onClose, ...other } = props;
+
+  return (
+    <DialogTitle {...other}>
+      {children}
+      {onClose ? (
+        <Button
+          aria-label="close"
+          onClick={onClose}
+          sx={{
+            position: 'absolute',
+            right: 26,
+            top: 27,
+          }}
+        >
+          <SvgIcon size={30} name={'close'} color={'white'} />
+        </Button>
+      ) : null}
+    </DialogTitle>
+  );
+};
+
+export function SpeedTestDialog(props) {
+  const { t } = useTranslation();
+  const { onClose, selectedValue, open, selectFocus } = props;
+  const { myLocalData, setSelectedCamera, selectedCamera, setSelectedMicrophone, selectedMicrophone, speedTestBeforeLogin, setSpeedTestBeforeLogin } = React.useContext(MediaSettingsContext);
+
+  const antmedia = React.useContext(AntmediaContext);
+  const { devices } = antmedia;
+
+  const theme = useTheme();
+  const fullScreen = useMediaQuery(theme.breakpoints.down('xs'));
+
+  const handleClose = (event, reason) => {
+    onClose(selectedValue);
+  };
+  function switchVideoMode(value) {
+    setSelectedCamera(value);
+    antmedia.switchVideoCameraCapture(myLocalData?.streamId, value);
+  }
+
+  function switchAudioMode(value) {
+    setSelectedMicrophone(value);
+    antmedia.switchAudioInputSource(myLocalData?.streamId, value);
+  }
+  React.useEffect(() => {
+    if (devices) {
+      const camera = devices.find(d => d.kind === 'videoinput');
+      const audio = devices.find(d => d.kind === 'audioinput');
+      if (camera && selectedCamera === '') setSelectedCamera(camera.deviceId);
+      if (audio && selectedMicrophone === '') setSelectedMicrophone(audio.deviceId);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [devices]);
```

Comment:
Forget about my last comment. You can improve this way better using a single `forEach` for almost all your iterations over `devices` 

```suggestion
    const [availableDevices, setAvailableDevices] = React.useState({audio: [], video: []})
    
    React.useEffect(() => {
      if (devices) {
        const availableVideo = [];
        const availableAudio = [];

        let camera = null;
        let audio = null;
        
        devices.forEach((d)=>{
          
          const menuObject = <MenuItem key={d.deviceId} value={d.deviceId}>
            {d.label}
          </MenuItem>;

          if(d.kind === 'videoinput'){
            
            // If no camera selected yet, set it
            if(!camera && selectedCamera === ''){
              camera = d
              setSelectedCamera(d.deviceId);              
            }
            
            availableAudio.push(menuObject);
          }else if(d.kind === 'audioinput'){
            // If no mic selected yet, set it 
            if(!audio && selectedMicrophone === ''){
              audio = d
              setSelectedMicrophone(d.deviceId);              
            }
            availableAudio.push(menuObject);
          }else{
            // Unknown device???? maybe does not exist, so you may get rid off this else-block :) 
          }
        });

        setAvailableDevices({
          audio: availableAudio, // Mics
          video: availableVideo // Cameras
        })
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [devices]);
```

Reply:
I'd also suggest o move it to a functions instead of using it on the useEffect.